File: C++/ps_baekjoon/11054/main.cpp
//baekjoon ps 11054

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main()
{
    ios_base::sync_with_stdio(false);
    int n;
    int res = 1;
    cin >> n;
    vector<int> arr(n);
    vector<int> dp1(n, 1);
    vector<int> dp2(n, 1);
    for (int i = 0; i < n; i++)
        cin >> arr[i];
    for (int i = 1; i < n; i++)
    {
        static int maxi;
        maxi = 1;
        for (int j = 0; j < i; j++)
            if (arr[i] > arr[j])
                maxi = max(maxi, dp1[j] + 1);
        dp1[i] = maxi;
    }
    for (int i = n - 1; i >= 0; i--)
    {
        static int maxi;
        maxi = 1;
        for (int j = n - 1; i < j; j--)
            if (arr[i] > arr[j])
                maxi = max(maxi, dp2[j] + 1);
        dp2[i] = maxi;
    }
    for (int i = 0; i < n; i++)
        res = max(res, dp1[i] + dp2[i]-1);
    cout << res;
    return 0;
}

File: Java/ps_baekjoon/11054/Main.java
import java.io.*;
import java.util.*;

public class Main {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    String readLine(){
        try{
            return br.readLine();
        }catch(Exception e){
            throw new RuntimeException(e);
        }
    }

    int n;
    int[] arr;
    int[] dp1, dp2;

    void solve(){
        n = Integer.parseInt(readLine());
        arr = Arrays.stream(readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        dp1 = new int[arr.length];
        dp2 = new int[arr.length];
        Arrays.fill(dp1, 1);
        Arrays.fill(dp2, 1);

        for(int i=0;i<dp1.length;++i){
            for(int j=0;j<i;++j){
                if(arr[j] < arr[i])
                    dp1[i] = Integer.max(dp1[i] , dp1[j] + 1);
            }
        }
        for(int i=dp1.length - 1;i>=0;--i){
            for(int j=dp1.length - 1;j>i;--j){
                if(arr[j] < arr[i])
                    dp2[i] = Integer.max(dp2[i], dp2[j] + 1);
            }
        }

        int maxi = 1;
        for(int i=0;i<arr.length;++i)
            maxi = Integer.max(maxi, dp1[i] + dp2[i] - 1);
        System.out.print(maxi);

    }
    public static void main(String[] args) throws IOException {
        new Main().solve();
    }
}